modin/engines/modin_dataframe.py

```diff
@@ -0,0 +1,406 @@
+        """
+        pass
+
+    def infer_types(self, columns_list):
+        """Determines the compatible type shared by all values in the specified columns,
+        and converts all values to that type.
+
+        Parameters
+        ----------
+            columns_list: list of strings
+                List of column labels to infer and induce types over.
+
+        Returns
+        -------
+        ModinDataframe
+            A new ModinDataframe with the inferred schema.
+        """
+        pass
+
+    def join(self, axis, condition, other, join_type):
+        """Joins this dataframe with the other.
+
+        Notes
+        -----
+            During the join, this dataframe is considered the left, while the other is
+                treated as the right.
+
+            Only inner joins, left outer, right outer, and full outer joins are currently supported.
+                Support for other join types (e.g. natural join) may be implemented in the future.
+
+        Parameters
+        ----------
+            axis: int
+                The axis to perform the join on.
+            condition: callable
+                Function that determines which rows should be joined. The condition can be a
+                simple equality, e.g. "left.col1 == right.col1" or can be arbitrarily complex.
+            other: ModinDataframe
+                The other data to join with, i.e. the right dataframe.
+            join_type: string
+                The type of join to perform.
+
+        Returns
+        -------
+        ModinDataframe
+            A new ModinDataframe that is the result of applying the specified join over the two
+            dataframes.
+        """
+        pass
+
+    def concat(self, axis, others):
+        """Appends the rows of identical column labels from multiple dataframes.
+
+        Notes
+        -----
+            The concat operator incurs fixed overheads, and so this algebra places no
+                limit to the number of dataframes that may be concatenated in this way.
+
+        Parameters
+        ----------
+            axis: int
+                The axis on which to perform the concatenation.
+            others: ModinDataframe or list of ModinDataframes
+                The other ModinDataframe(s) to concatenate.
+
+        Returns
+        -------
+        ModinDataframe
+            A new ModinDataframe that is the result of concatenating the dataframes over the
+            specified axis.
+        """
+        pass
+
+    def transpose(self):
+        """Swaps the row and column axes.
+
+        Notes
+        -----
+            Transposing a dataframe is expensive, and thus, while the axes are swapped
+                logically immediately, the physical swap does not occur until absolutely necessary,
+                which helps motivate the axis argument to the other operators in this algebra.
+
+            This operation explicitly manipulates dataframe metadata.
```

Comment:
I added this note to any function that operated on metadata - do we want to keep it?

Reply:
I don't think it's necessarily useful.